Revise modular pipelines docs

DimedS opened the pull request:
## Description
Revised the [pipeline](https://github.com/kedro-org/kedro/blob/1998-revise-the-modular-pipelines-docs/docs/source/nodes_and_pipelines/pipeline_introduction.md) and [modular-pipelines](https://github.com/kedro-org/kedro/blob/1998-revise-the-modular-pipelines-docs/docs/source/nodes_and_pipelines/modular_pipelines.md) documentation sections, introduced [namespaces](https://github.com/kedro-org/kedro/blob/1998-revise-the-modular-pipelines-docs/docs/source/nodes_and_pipelines/namespaces.md) docs section.

Main Changes:

1. Sorted topics within these sections: simple usage is now in the pipeline introduction, while pipeline reusage is covered in the advanced-level second document.
2. Added a table of contents with links at the top of the pipeline introduction.
3. Also added a `Pipeline creation structure` section, which describes how to create a pipeline after using the `kedro pipeline create` command, what to include in` nodes.py` and `pipeline.py`, and why, as well as other available options.
4. Grouped key pipeline methods and operations.
5. Simplified the pipeline reuse documentation to focus more on practical implementation.
6. Reduced the size of pictures by 60% as they were too large and not useful.


## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

merelcht commented:
Rendered docs for anyone else reviewing: https://kedro--3948.org.readthedocs.build/en/3948/

merelcht reviewed (dismissed):
Thanks so much for taking on this work @DimedS, you've clearly put a lot of thought in restructuring the docs and I really appreciate the effort. I've left comments in the specific pages but I think on a higher level the flow isn't quite right.

I found that the new structure seems to disrupt the logical flow of information. Sections don't seamlessly transition into one another, making it challenging to follow along. As a result, I'm concerned that the overall clarity and effectiveness of the documentation may have been compromised.

To address this, I propose revisiting the organisation of the content to ensure that each section logically leads into the next, creating a smoother reading experience. 

For example, the sections on [how to create a pipeline](https://kedro--3948.org.readthedocs.build/en/3948/nodes_and_pipelines/pipeline_introduction.html#how-to-create-a-simple-pipeline) and [create a new blank pipeline with the kedro pipeline create command](https://kedro--3948.org.readthedocs.build/en/3948/nodes_and_pipelines/pipeline_introduction.html#create-a-new-blank-pipeline-with-the-kedro-pipeline-create-command) seem very separate and then in the section on pipeline creation structure you jump back again to the example used in "how to create a simple pipeline".

Additionally, we may need to reconsider the level of detail provided in certain sections to strike the right balance between comprehensive coverage and concise readability. 

I really like the rephrasing of "modular pipelines" to "reusable pipelines", but from just reading the doc, I don't feel like I'd be able to apply the information in practice. I have always hated that cooking pipeline example and if we change anything I think it has to be that example. In the ticket for this task [Revise the modular pipelines documentation to improve clarity](https://github.com/kedro-org/kedro/issues/1998) Jo mentioned "I'd like to see it more like a tutorial that is based on the spaceflights starter (extends it)." and I totally agree with that. 


I'm aware I haven't been involved much in this task, so if any of this goes against what you and @astrojuanlu have discussed, feel free to dismiss it!

ankatiyar commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -1,26 +1,30 @@
 # Pipelines
 
-We previously introduced [Nodes](./nodes.md) as building blocks that represent tasks, and can be combined in a pipeline to build your workflow. A pipeline organises the dependencies and execution order of your collection of nodes, and connects inputs and outputs while keeping your code modular. The pipeline resolves dependencies to determine the node execution order, and does *not* necessarily ru ...
+A pipeline is a set of [Nodes](./nodes.md) that will be automatically resolved after pipeline creation to determine the node execution order based on outputs and inputs of each node.
```
Not sure if this sentence reads very clear to me. I don't really mind the introduction from before but simplified just a little bit?

ankatiyar commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -1,26 +1,30 @@
 
-## How to build a pipeline
+Most important about kedro pipelines:
+- [How to create a simple pipeline](#how-to-create-a-simple-pipeline)
+- [Create a new blank pipeline with kedro pipeline create command](#create-a-new-blank-pipeline-with-the-kedro-pipeline-create-command)
+- [Pipeline creation structure](#pipeline-creation-structure)
+- [Key Pipeline methods and operations](#key-pipeline-methods-and-operations)
+- [Bad pipelines - rules](#bad-pipelines)
+- [Custom new pipeline templates](#custom-new-pipeline-templates)
```
I think these headings could be all "how to" headlines like we have in configuration docs. https://docs.kedro.org/en/stable/configuration/advanced_configuration.html#advanced-configuration-for-kedro-projects

ankatiyar commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
````diff
@@ -34,12 +38,112 @@ variance_pipeline = pipeline(
 )
 ```
-</details>
 
````
A little explanation here about how Kedro can figure out the order of execution of these nodes based on the inputs and outputs of the nodes in a pipeline.

ankatiyar commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -34,12 +38,112 @@ variance_pipeline = pipeline(
-<summary><b>Click to expand</b></summary>
+## Create a new blank pipeline with the `kedro pipeline create` command
+
+When managing your Kedro project, we recommend organising your pipelines separately to achieve modularity. This approach allows for easy copying and reuse of pipelines within and between projects. Simply put: one pipeline, one folder. To help with that, you can use the following command (the pipeline name must adhere to [Python convention](https://realpython.com/python-pep8/#naming-conventions)):
```
This also reads a little bit unclear to me. I think this part could explain the idea behind dividing tasks into pipelines. For eg, a project usually contains a lot of tasks and we recommend grouping tasks that frequently are executed together into individual pipelines. These pipelines should ideally be organised in separate folders to encourage reuse and modularity.

ankatiyar commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
````diff
@@ -34,12 +38,112 @@ variance_pipeline = pipeline(
+
+```
+
+If you want to do the reverse and remove that pipeline, you can use `kedro pipeline delete <pipeline_name>` to do so or have a look to all [project-specific CLI commands list](../development/commands_reference.md#kedro-commands).
````
```suggestion
If you want to delete an existing pipeline, you can use `kedro pipeline delete <pipeline_name>` to do so. 
```

DimedS commented:
Thank you very much for the comprehensive review, @merelcht! I agree that the current flow is not ideal. Let's brainstorm how we can restructure it. My main idea was to split the content into two docs:
1. Basic topics that everyone should know (including how to use the Kedro pipeline create command and our recommended practices for writing pipeline code).
2. Advanced topics on pipeline reuse, including namespacing and overwriting.
I also agree that the cooking example isn't great. I'd be happy to replace it with a more realistic scenario, like a spaceflights extension.
Let's wait for some more feedback and then think how to refine the structure!

ankatiyar commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -34,12 +38,112 @@ variance_pipeline = pipeline(
+    namespace = "", # Optional
+    inputs = {}, # Optional
+    outputs = {}, # Optional
+    parameters = {}, # Optional
```
Might be missing a closing paranthesis?

ankatiyar commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
````diff
@@ -60,36 +164,11 @@ variance_node
 ```
-</details>
-
-### How to tag a pipeline
````
Not sure why this section is removed from the page?

ankatiyar commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,156 +1,57 @@
-# Modular pipelines
+# Pipeline reusage with namespaces
```
Maybe "Reuse pipelines using namespaces"? Reusage doesn't sound right :P

ankatiyar commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,156 +1,57 @@
-            └── {{pipeline_name}}      <-- Pipeline-specific tests
-                ├── __init__.py
-                └── test_pipeline.py
+If you need to try different options for training your model, constantly overwriting your outputs for each new pipeline can become tedious. Each model's results (outputs) should be saved in new objects, which can be annoying to manage manually. Namespaces are a perfect solution for this. By setting a `namespace="namespace_name"` parameter for each new pipeline (based on an existing pipeline), you ...
```
```suggestion
If you need to try different options for training your model, constantly overwriting your outputs for each new pipeline can become tedious. Each model's results (outputs) should be saved in new objects, which can be annoying to manage manually. Namespaces are a perfect solution for this. By setting a `namespace="namespace_name"` parameter for each new pipeline (based on an existing pipeline), you achieve automatic output and node isolation. This is done by adding a `namespace_name` prefix to each pipeline node and output dataset. Your pipeline creation code will look like this:
```
Might also be worth showing the example of adding the namespace prefix in the catalog and parameter files?

ankatiyar commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,156 +1,57 @@
+## How to share your pipelines
 
-### Custom templates
+Pipelines are shareable between Kedro codebases via [micro-packaging](micro_packaging.md), but you must follow a couple of rules to ensure portability:
```
Micropackaging is going to be deprecated so perhaps it's worth adding alternate instructions on how to do this?

DimedS commented:
@merelcht, @ankatiyar, thank you for your reviews! I have addressed most of your comments. I would appreciate any advice on micropackaging and providing a clearer explanation of namespaces. Additionally, I did not modify the cookie examples for modular pipelines; let's confirm any changes regarding that with @astrojuanlu and Jo.

astrojuanlu commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -252,3 +249,131 @@ Nodes that are created with input or output names that contain `.` risk a discon
+
+## How to create a new blank pipeline using the `kedro pipeline create` command
+
+When managing your Kedro project, we recommend grouping related tasks into individual pipelines to achieve modularity. A project typically contains many tasks, and organising frequently executed tasks together into separate pipelines helps maintain order and efficiency. Each pipeline should ideally be organised in its own folder, promoting easy copying and reuse within and between projects. Simpl ...
```
From here I'd direct readers to the next page, explaining Modular Pipelines (see review comment)

```suggestion
When managing your Kedro project, we recommend grouping related tasks into individual pipelines to achieve modularity. A project typically contains many tasks, and organising frequently executed tasks together into separate pipelines helps maintain order and efficiency. Each pipeline should ideally be organised in its own folder, promoting easy copying and reuse within and between projects. Simply put: one pipeline, one folder. To assist with this, Kedro introduces the concept of [Modular Pipelines](./modular_pipelines), which are described in the next section.
```

astrojuanlu commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,156 +1,66 @@
-# Modular pipelines
+# Reuse pipelines with namespaces
```
This `modular_pipelines.md` could be `namespaces.md`, and then `modular_pipelines.md` should be for the last "how-to" sections of `pipeline_introduction.md` (see review comment)

astrojuanlu commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
````diff
@@ -1,156 +1,66 @@
-* It also allows you to simplify both your mental models, and Kedro-Viz visualisations via `namespaces`.
-
-```python
-from kedro.pipeline.modular_pipeline import pipeline
````
Happy to see this line go away 😄

astrojuanlu reviewed (commented):
Gave this a first pass. Fantastic job @DimedS 👏🏼 

Some high level thoughts:

- I really like all the "how-tos" in https://kedro--3948.org.readthedocs.build/en/3948/nodes_and_pipelines/pipeline_introduction.html
- However, from https://kedro--3948.org.readthedocs.build/en/3948/nodes_and_pipelines/pipeline_introduction.html#how-to-create-a-new-blank-pipeline-using-the-kedro-pipeline-create-command onwards, we're talking about Modular Pipelines, because it describes the directory structure that wraps `Pipeline` objects (and here the use of "Modular" is correct!).
- And then, the document called "Reuse pipelines with namespaces" goes on to explain mostly namespaces, which I think is good 👍🏼 but it's still named `modular_pipelines.md`

So I only have 1 major about the structure, which is having "Modular Pipelines" still have a page on its own. Something like:

<img width="281" alt="image" src="https://github.com/kedro-org/kedro/assets/316517/b2f65998-6b9d-4780-b122-480616afad3a">

noklam commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,156 +1,66 @@
+# Reuse pipelines with namespaces
 
-## What are modular pipelines?
+In many typical Kedro projects, a single (“main”) pipeline increases in complexity as the project evolves. To keep your project fit for purpose, we recommend you separate your code into different pipelines (modules) that are logically isolated and can be reused.
```
I don't think this is the purpose of namespace. Even without namespace, users can have multiple pipeline. i.e.

- pipelines
  - data_science
  - data_processing

noklam commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,156 +1,66 @@
+## How to reuse your pipelines
 
-Modular pipelines allow you to instantiate pipelines multiple times, while allowing the user to override inputs/outputs/parameters. They are reusable within the same codebase, and shareable across projects via [micro-packaging](micro_packaging.md). This is the modern way to use Kedro, and will change the way you think about your pipelines.
+If you want to create a new pipeline that performs the same tasks as your existing pipeline (e.g., `data_science`), you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](pipeline_introduction.md#how-to-structure-your-pipeline-creation). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline creation code sho ...
```
```suggestion
If you want to create a new pipeline that performs the same tasks as your existing pipeline (e.g., `data_science`), you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](pipeline_introduction.md#how-to-structure-your-pipeline-creation). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline creation code should look like this:
```

Similar tasks with different inputs/outputs?

noklam commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,156 +1,66 @@
+def create_new_pipeline(**kwargs) -> Pipeline:
+    return pipeline(
+    [data_science], # Name of an existing pipeline
+    inputs = {"old_input__df_name" : "new_input_df_name"},
```
```suggestion
    inputs = {"old_input_df_name" : "new_input_df_name"},
```

noklam commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -2,25 +2,29 @@
+To benefit from Kedro's automatic dependency resolution, you can chain your nodes into a {py:class}`~kedro.pipeline.Pipeline`, which is a list of nodes that use a shared set of variables. That class could be created using {py:class}`~kedro.pipeline.modular_pipeline.pipeline` method, based on nodes or other pipelines (in which case all nodes from that pipeline will be used).
 
-## How to build a pipeline
+- [How to create a simple pipeline](#how-to-create-a-simple-pipeline)
```
I actually think this is a good suggestion to keep the original wording.

noklam commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -34,12 +38,24 @@ variance_pipeline = pipeline(
 
-<details>
-<summary><b>Click to expand</b></summary>
+Kedro's dependency resolution ensures that each node runs only after its required inputs are available from the outputs of previous nodes. This way, the nodes are executed in the correct order automatically, based on the defined dependencies.
```
Maybe it helps to print `pipline.nodes`  which is toposorted already?

DimedS replied:
It's the section below titled "How to receive information about the nodes in a pipeline" that shows that output, so I don't think it's a good idea to duplicate it on the same page.

noklam reviewed (commented):
I left some minor comments but I thought it's best to review the high-level structure first before going into minor details.

Even our own docs is confusing, we have a page title "Reuse pipeline with namespace" while having the URL as "modular_pipelines.html". They are really different thing.
https://kedro--3948.org.readthedocs.build/en/3948/nodes_and_pipelines/modular_pipelines.html

Question:
- With micro-packging deprecating, do we fade out the "modular pipeline" concept too? My answer is no since `kedro pipeline create` will still use it and in general is a nice idea to keep your pipeline "modular" in terms of configuration, or even dependencies. (So it's easier to run in Orchestrator)

With that in mind, I think we should explains clearly "What is a modular pipeline", it is mainly about the structure, organisation of config/dependencies, what are the benefits etc. 

"Modular pipeline" is a concept, then "Namespace" is a Kedro feature that helps you to implement pipeline in a modular way.

astrojuanlu commented:
> With micro-packging deprecating, do we fade out the "modular pipeline" concept too? My answer is no since kedro pipeline create will still use it

Agreed, we should keep the term

DimedS commented:
@astrojuanlu, @noklam, thank you for your comments. I have merged modular pipelines with pipeline objects on one page to simplify things for beginner users. I felt that it wasn't essential for them to understand these as two conceptual objects. Instead, they can simply use "kedro pipeline create," understand how to fill the template, and get started. However, I am not happy with the current structure because it is too long, and the "kedro pipeline create" section is at the end. Therefore, @astrojuanlu, I fully support your proposed structure and will create a separate page for modular pipelines with a clear explanation of the folder structure.

DimedS commented:
I have changed the structure according to @astrojuanlu's proposal and addressed other minor comments. I am now re-requesting reviews.

astrojuanlu commented on docs/source/faq/faq.md:
```diff
@@ -52,7 +52,7 @@ This is a growing set of technical FAQs. The [product FAQs on the Kedro website]
 ## Nodes and pipelines
 
-* [How do I create a modular pipeline](../nodes_and_pipelines/modular_pipelines.md#how-do-i-create-a-modular-pipeline)?
+* [How can I reuse my pipelines](../nodes_and_pipelines/modular_pipelines.md)?
```
Outdated link

astrojuanlu commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -34,12 +44,17 @@ variance_pipeline = pipeline(
+3. The third node computes the mean sum of squares (mean_sos) using `xs` and `n`, and outputs it as `m2`.
+4. The fourth node calculates the variance using the mean (`m`) and mean sum of squares (`m2`), and outputs it as `v`.
+
+Kedro's dependency resolution ensures that each node runs only after its required inputs are available from the outputs of previous nodes. This way, the nodes are executed in the correct order automatically, based on the defined dependencies.
```
```suggestion
Kedro's dependency resolution algorithm ensures that each node runs only after its required inputs are available from the outputs of previous nodes. This way, the nodes are executed in the correct order automatically, based on the defined dependencies.
```

?

astrojuanlu commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
````diff
@@ -34,12 +44,17 @@ variance_pipeline = pipeline(
+
+## How to use `describe` to discover what nodes are part of the pipeline
 
 ```python
````
Maybe one line explaining what this is, even if it's short

astrojuanlu commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -58,7 +28,7 @@ Running the `kedro pipeline create` command adds boilerplate folders and files f
     │   └── pipelines
     │       ├── __init__.py
-    │       └── {{pipeline_name}}      <-- This folder defines the modular pipeline
+    │       └── {{pipeline_name}}      <-- This folder defines the pipeline
```
Maybe keep "modular" here

astrojuanlu commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
````diff
@@ -72,11 +42,71 @@ Running the `kedro pipeline create` command adds boilerplate folders and files f
-</details>
+If you want to delete an existing pipeline, you can use `kedro pipeline delete <pipeline_name>` to do so.
+```{note}
+To see the full list of available CLI options, you can run `kedro pipeline create --help` for more information.
````
```suggestion
To see the full list of available CLI options, you can run `kedro pipeline create --help`.
```

?

astrojuanlu commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
````diff
@@ -72,11 +42,71 @@ Running the `kedro pipeline create` command adds boilerplate folders and files f
+    return pipeline([])
+```
+Here, you are creating a function that returns a `Pipeline` class instance with the help of the `pipeline` function. We recommend putting your pipeline creation code into a function because it allows to achieve:
+- Modularity: you can easily reuse this logic across different parts of your project.
````
I'd maybe skip this bullet list, or I'd move it to the top of the page, where the concept is explained and justified

DimedS replied:
I changed them with explanation that function should be named `create_pipeline()` for auto-discovery.

astrojuanlu commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,221 @@
+```python
+def create_new_pipeline(**kwargs) -> Pipeline:
+    return pipeline(
+    [data_science], # Name of an existing pipeline
````
Where does `data_science` come from in this example?

DimedS replied:
I refer to a `data_science` pipeline from the Spaceflights starter tutorial. In this example, it's clearer why a user might want to reuse parts of the pipeline in a business case, compared to an example using medians and means. I've included links to the tutorial.

astrojuanlu commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,221 @@
+
+## How to share your pipelines
+
+Pipelines are shareable between Kedro codebases via [micro-packaging](micro_packaging.md), but you must follow a couple of rules to ensure portability:
```
We need to add a warning here clarifying that micro packaging is deprecated and slated for removal.

Also, if anything, shouldn't this section be in `modular_pipelines.py`?

astrojuanlu reviewed (commented):
Gave another pass, left some comments. We're on the right path!

astrojuanlu reviewed (commented):
Also wondering why the images are smaller now? 😅

DimedS commented:
> Also wondering why the images are smaller now? 😅

I resized them to smaller ones because previously I saw them (on the screen below) in a very weird way:
<img width="1602" alt="Screenshot 2024-06-24 at 11 33 13" src="https://github.com/kedro-org/kedro/assets/40151847/8a538615-f1c9-4e2b-8ba2-9b415b746867">

DimedS commented:
> Gave another pass, left some comments. We're on the right path!

Thank you for your additional comments, @astrojuanlu. I agree with them and have modified the docs accordingly. It would be great if you could take another look at them.

astrojuanlu commented:
> I resized them to smaller ones because previously I saw them (on the screen below) in a very weird way:

Oh I see. Can we control the size on the docs instead? Something like

````
```{image} ../meta/images/complete-mlflow-tracking-kedro-mlflow.png
:alt: Complete MLflow tracking with kedro-mlflow
:width: 80%
:align: center
```
````

(and we revert the size changes of the images)

astrojuanlu commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -88,10 +115,9 @@ A template folder passed to `kedro pipeline create` using the `--template` argum
 separate folder and pointing to them with the `--template` flag.
 
-#### Creating custom templates
+### Creating custom templates
```
```suggestion
### Creating custom pipeline templates
```

astrojuanlu reviewed (approved):
From a structure and flow perspective I think this is already a massive improvement over what we had now 💯

DimedS commented:
> > I resized them to smaller ones because previously I saw them (on the screen below) in a very weird way:
> 
> Oh I see. Can we control the size on the docs instead? Something like
> 
> ```
> ```{image} ../meta/images/complete-mlflow-tracking-kedro-mlflow.png
> :alt: Complete MLflow tracking with kedro-mlflow
> :width: 80%
> :align: center
> ```
> ```
> 
> (and we revert the size changes of the images)

Why do you think it's better to download a larger image and resize it every time a user opens a page, instead of resizing it once to produce a smaller image?

astrojuanlu commented:
My thinking was that it would be good to retain the original high-resolution images in case somebody has a high density screen, or wants to zoom-in for whatever reason, or we can fully use the high-res version if/when we redesign how images are placed (#2643) or the whole docs (https://github.com/kedro-org/kedro-sphinx-theme/issues/5)

No strong objection though

noklam reviewed (commented):
I like the new structure better and agree this is in the right direction.

My main comment are with `namespaces.md`:
- I found the example slightly confusing, as we switch between different examples. `i.e. old_input_df_name, `alpha,`beta`, `grill`,`defrost`
- It's not clear from the documentation what `namespace` actually do other than `This is done by adding a namespace_name prefix to each pipeline node and output dataset. Your pipeline creation code will look like this`. This is usually where the confusion comes from. I'd first show what namespace does do a pipeline, then explains how to use the argument like[` parameters`](https://kedro--3948.org.readthedocs.build/en/3948/nodes_and_pipelines/namespaces.html#example-how-to-reuse-a-pipeline-with-different-parameters) to stitch pipelines
- I find the example of ["disconnect pipeline"](https://kedro--3948.org.readthedocs.build/en/3948/nodes_and_pipelines/namespaces.html#example-combining-disconnected-pipelines) confusing, it shows you can override data with `inputs`, `outputs` which feels duplicate with the example later.

stichbury commented on docs/source/faq/faq.md:
```diff
@@ -52,8 +52,8 @@ This is a growing set of technical FAQs. The [product FAQs on the Kedro website]
 
-* [How do I create a modular pipeline](../nodes_and_pipelines/modular_pipelines.md#how-do-i-create-a-modular-pipeline)?
-
+* [How can I create a new blank pipeline using the `kedro pipeline create` command](../nodes_and_pipelines/modular_pipelines.md#how-to-create-a-new-blank-pipeline-using-the-kedro-pipeline-create-command)?
```
```suggestion
* [How can I create a new blank pipeline](../nodes_and_pipelines/modular_pipelines.md#how-to-create-a-new-blank-pipeline-using-the-kedro-pipeline-create-command)?
```

stichbury replied:
Do you need the "using the `kedro pipeline create` command" part of this sentence?

stichbury commented on docs/source/faq/faq.md:
```diff
@@ -52,8 +52,8 @@ This is a growing set of technical FAQs. The [product FAQs on the Kedro website]
-* [How do I create a modular pipeline](../nodes_and_pipelines/modular_pipelines.md#how-do-i-create-a-modular-pipeline)?
-
+* [How can I create a new blank pipeline using the `kedro pipeline create` command](../nodes_and_pipelines/modular_pipelines.md#how-to-create-a-new-blank-pipeline-using-the-kedro-pipeline-create-command)?
+* [How can I reuse my pipelines with namespaces](../nodes_and_pipelines/namespaces.md)?
```
```suggestion
* [How can I reuse my pipelines](../nodes_and_pipelines/namespaces.md)?
```

stichbury replied:
Same here really. If there are multiple ways and you're solving for a particular way, e.g. with namespaces, maybe it is OK, but otherwise I'd avoid making it so long.

stichbury commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,212 @@
+
+## How to reuse your pipelines
+
+If you want to create a new pipeline that performs similar tasks with different inputs/outputs/parameters as your existing pipeline (e.g., `data_science` pipeline from our [Spaceflights tutorial](../tutorial/add_another_pipeline.md#data-science-pipeline)), you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](modular_pipelines.md#how-to-stru ...
```
>If you want to create a new pipeline that performs similar tasks with different inputs/outputs/parameters as your existing pipeline (e.g., `data_science` pipeline from our [Spaceflights tutorial](../tutorial/add_another_pipeline.md#data-science-pipeline)), you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](modular_pipelines.md#how-to-structure-your-pipeline-creation). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline creation code should look like this:


I found this hard to parse.  It's one very long sentence with two links, plus some parenthesis. I would split it and consider whether you need all the links.

stichbury commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,212 @@
+```
+This means you can easily create multiple pipelines based on the `data_science` pipeline to test different approaches with various input datasets and model training parameters.
+
+## What is a Namespace
````
Is there a reason for capitalising namespace?

stichbury commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,53 +1,23 @@
+- [How to share your pipelines](#how-to-share-your-pipelines)
 
-Modular pipelines allow you to instantiate pipelines multiple times, while allowing the user to override inputs/outputs/parameters. They are reusable within the same codebase, and shareable across projects via [micro-packaging](micro_packaging.md). This is the modern way to use Kedro, and will change the way you think about your pipelines.
+## How to create a new blank pipeline using the `kedro pipeline create` command
```
Long subheads like this look very awkward in the left hand table of contents. Is it necessary to have the full title or could it just be "How to create a new blank pipeline" ?

stichbury commented on docs/source/nodes_and_pipelines/modular_pipelines.md:
```diff
@@ -1,53 +1,23 @@
-## How do I create a modular pipeline?
-
-You can use a [project-specific CLI command](../development/commands_reference.md#kedro-commands) to create a modular pipeline. The pipeline name must adhere to [Python convention](https://realpython.com/python-pep8/#naming-conventions).
+To assist with creating a new pipeline that supports the modularity concept, you can use the following command (the pipeline name must adhere to [Python naming conventions](https://realpython.com/python-pep8/#naming-conventions)):
```
> To assist with creating a new pipeline that supports the modularity concept, you can use the following command (the pipeline name must adhere to [Python naming conventions](https://realpython.com/python-pep8/#naming-conventions)):

This is a verbose way to basically just say:
```suggestion
 To create a new modular pipeline, use the following:
```

stichbury reviewed (commented):
Left a few comments but the structure looks OK to me and the team are far better placed to comment on that anyway. Nice work!

Can I suggest that you look through the Vale warnings and address them because at present the main issue is that it's wordy and using lexicon that we try to avoid (e.g. "easily" is not something we say because it may be easy for us, but isn't for everyone). 

The Vale checks are there to help improve the readability and I feel like it suffers slightly from not following the Vale style.

DimedS commented:
Thank you for your comments @noklam, @stichbury, @merelcht. I have revised the [namespaces](https://github.com/kedro-org/kedro/blob/1998-revise-the-modular-pipelines-docs/docs/source/nodes_and_pipelines/namespaces.md) section by adding examples based on the spaceflights-pandas starter. I hope this provides clarity on how namespaces work and how to use them effectively. Could you please re-review that section?

merelcht commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,345 @@
+> Tip: You can use `kedro run --namespace = namespace_name` to run only the specific namespace
+
+
+### How make all pipelines in this example fully namespaced
```
```suggestion
### How to namespace all pipelines in a project
```

merelcht commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,345 @@
+        namespace="data_processing",
+        inputs={"companies", "shuttles", "reviews"},
+```
+Modify the `data_science` pipeline by adding namespace and inputs the same way as it was in `data_science_2` pipeline, but the input will be prefixed with `data_processing.`, because we added a namespace for data processing in the previous step:
````
```suggestion
Modify the `data_science` pipeline by adding namespace and inputs in the same way as it was done in `data_science_2` pipeline, but the input will be prefixed with `data_processing.`, because we added a namespace for data processing in the previous step:
```

merelcht commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,345 @@
+inputs={"data_processing.model_input_table"},
+```
+
+After executing the pipeline with `kedro run`, the visualization with `kedro viz run` after collapsing will look like this:
````
```suggestion
After executing the pipeline with `kedro run`, the visualisation with `kedro viz run` after collapsing will look like this:
```

merelcht reviewed (commented):
The structure looks great now! I left some comments on language and agree with @stichbury that it's worth checking the Vale suggestions.

One larger thing I noticed is that on the new `namespaces.md` page there's a switch in person in the different section. In some parts it uses the "you" person, in others it uses "we". That should be made consistent to make the full page easier to read.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,336 @@
+```python
+def create_new_pipeline(**kwargs) -> Pipeline:
+    return pipeline(
+    [existing_pipeline], # Name of the existing pipeline
````
Why is this`[existing_pipeline]` instead of  `existing_pipeline`? Also the comment is a bit inaccurate: it's not `# Name of the existing pipeline` but rather a `Pipeline` object itself.

yury-fedotov replied:
I think the example would be more realistic if `existing_pipeline` had a simple but realistic job and clear name. E.g. like `cook_pipeline` in old example.

DimedS replied:
I fixed the comment and removed the brackets, but I believe `existing_pipeline` is a sufficiently descriptive name to define reusability. A more detailed example is provided a bit further down, based on the spaceflights starter.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,336 @@
+def create_new_pipeline(**kwargs) -> Pipeline:
+    return pipeline(
+    [existing_pipeline], # Name of the existing pipeline
+    inputs = {"old_input_df_name" : "new_input_df_name"},  # Mapping old input to new input
+    outputs = {"old_output_df_name" : "new_output_df_name"},  # Mapping old output to new output
+    parameters = {"params: model_options": "params: new_model_options"},  # Updating parameters
```
I think the `old` and `new` dataset names do not reflect what modular pipelines really do.

A mental model while working with modular pipelines and namespaces is not that you're replacing something **old** with something **new**. It is that you're using something **specific** instead of something **abstract**.

Previous example with cooking food I think reflects it better because it shows how `"grilled_veg"` which is an output of `cook_pipeline` is in one case `"breakfast_food"` and in the other it is `"lunch_food"`

yury-fedotov replied:
Also the block inside `pipeline()` should be indented I believe

DimedS replied:
I fixed the indentation and updated the comments. I believe the idea is now also clearly explained within the example—showing how to create different data science pipelines based on the base_data_science pipeline.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,336 @@
+
+## How to reuse your pipelines
+
+If you want to create a new pipeline that performs similar tasks with different inputs/outputs/parameters as your `existing_pipeline`, you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](modular_pipelines.md#how-to-structure-your-pipeline-creation). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline cr ...
```
```suggestion
If you want to create a new pipeline that performs similar tasks with different inputs/outputs/parameters as your existing_pipeline, you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](modular_pipelines.md#how-to-structure-your-pipeline-creation). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline creation code should look like this:
```

yury-fedotov replied:
Suggest splitting it to multiple lines to simplify representation in GitHub.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,336 @@
+If you want to create a new pipeline that performs similar tasks with different inputs/outputs/parameters as your `existing_pipeline`, you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](modular_pipelines.md#how-to-structure-your-pipeline-creation). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline cr ...
+
+```python
+def create_new_pipeline(**kwargs) -> Pipeline:
````
I think this PR overall uses less specific names than the current docs, which at least to me is an impact on docs quality. Compare:

`create_new_pipeline`

to:

```python
final_pipeline = (
    cook_breakfast_pipeline
    + eat_breakfast_pipeline
    + cook_lunch_pipeline
    + eat_lunch_pipeline
)
```

DimedS replied:
I updated it to` create_pipeline()`. If I understood correctly, the name `create_pipeline()` is important (although it is not specific) because otherwise, the pipeline will not be autodiscovered and executed during `kedro new` (see [Kedro documentation](https://docs.kedro.org/en/stable/nodes_and_pipelines/pipeline_registry.html)). If you see any other examples with less specific names, please let me know and I would be happy to fix them as well.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,336 @@
+                func=split_data,
+                inputs=["model_input_table", "params:model_options"],
+                outputs=["X_train", "X_test", "y_train", "y_test"],
+                name="split_data_node",
```
And similarly in other nodes, because in all the places this `name` is used, it is clear that it's a `node`.
```suggestion
                name="split_data",
```

DimedS replied:
I would prefer to leave this as it is because it's currently how it's written in our starters, and I don't want to confuse users with different naming since our example is based on that starter.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,336 @@
+    )  # Creating a base data science pipeline that will be reused with different model training parameters
+
+
+def create_pipeline(**kwargs) -> Pipeline:
+    return pipeline(
+        [base_data_science],  # Creating a new data_science pipeline based on base_data_science pipeline
+        parameters={"params:model_options": "params:model_options_1"},  # Using a new set of parameters to train model
+    )
```
Why is it not using `namespace`? Isn't this page supposed to be about that?

DimedS replied:
In my opinion, the point of the first section titled "How to reuse your pipelines" is to show that you can reuse your pipelines without needing namespaces. However, by adding namespaces, you can reuse your pipelines multiple times, with namespaces providing the necessary support. So my point is: reuse is the primary idea, and namespaces are an enhancement.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,336 @@
+
+A namespace is a way to isolate nodes, inputs, outputs, and parameters inside your pipeline. If you put `namespace="namespace_name"` attribute inside the `pipeline()` creation function, it will add the `namespace_name.` prefix to all nodes, inputs, outputs, and parameters inside your new pipeline.
+
+Let's extend our previous example and try to reuse the `base_data_science` pipeline one more time by creating another pipeline based on it. First, we should use the `kedro pipeline create` command to create a new blank pipeline named `data_science_2`:
```
Ok this is very very unclear. Why to reuse an existing pipeline that's designed to be modular you're running `kedro pipeline create`? Isn't a more intuitive flow to define this abstract, unregistered pipeline once and then initialize it's actual instances to be registed within the same `project/pipelines/pipeline` folder?

DimedS replied:
Thank you for that proposal, @yury-fedotov. When I wrote that manual, I also thought about how I can reuse a pipeline created with `kedro pipeline create` command since the pipeline object is created inside the function. Should I create an additional pipeline with the `kedro pipeline create` command to reuse the first one? I believe it's important to provide our users with best practices, and I would be happy to hear more opinions on that. Maybe @astrojuanlu and @merelcht have some insights.

merelcht replied:
From my point of you it doesn't matter how you create the pipelines, what's important is how you re-use them in the end. It's up to the user if they want the pipelines to be in the same folder or separated out (which is what happens when you do `kedro pipeline create`).

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,336 @@
+def create_pipeline(**kwargs) -> Pipeline:
+    return pipeline(
+        [base_data_science], # Creating a new data_science_2 pipeline based on base_data_science pipeline
+        parameters={"params:model_options": "params:model_options_2"}, # Using a new set of parameters to train model
+        namespace = "ds_2", # With that namespace, "ds_2." prefix will be added to inputs, outputs, params, and node names
+        inputs={"model_input_table"}, # Inputs remain the same, without namespace prefix
```
This order of arguments doesn't make sense. `namespace` should be either below all `inputs/outputs/parameters` or after. Not in the middle. They are indeed kwargs and order doesn't matter but for readability and user understanding it does matter here I think.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,336 @@
+```python
+def create_pipeline(**kwargs) -> Pipeline:
+    return pipeline(
+        [base_data_science],
````
Why is this a list?
```suggestion
        base_data_science,
```

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,336 @@
+
+![namespaces collapsed all](../meta/images/namespaces_collapsed_all.png)
+
+## Providing pipeline specific dependencies
+
+* A pipeline **might** have external dependencies specified in a local `requirements.txt` file.
+* Pipeline specific dependencies are scooped up during the [micro-packaging](micro_packaging.md) process.
+* These dependencies need to be manually installed using `pip`:
+```bash
+pip install -r requirements.txt
+```
````
How is this at all related to namespaces?

DimedS replied:
Fixed, I moved it to `modular_pipelines.md`

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,336 @@
+
+## Example: Combining disconnected pipelines
+
+Sometimes two pipelines must be connected, but do not share any catalog dependencies. In this example, there is a `lunch_pipeline`, which makes us lunch. The 'verbs', `defrost` and `eat`, are Python functions and the inputs/outputs are food at different points of the process (`frozen`, `thawed` and `food`).
```
`thawed` is not present in the example at all

DimedS replied:
Decided to remove the cooking examples.

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,336 @@
+```{note}
+The set of overriding inputs and outputs must be a subset of the reused pipeline's "free" inputs and outputs, respectively. A free input is an input that isn't generated by a node in the pipeline, while a free output is an output that isn't consumed by a node in the pipeline. {py:meth}`Pipeline.inputs() <kedro.pipeline.Pipeline.inputs>` and {py:meth}`Pipeline.outputs() <kedro.pipeline.Pipeline.ou ...
+```
+
````
Why do we need both cooking and data science example to demonstrate how namespaces work?

DimedS replied:
I agree with that. We want to move away from the cooking example and shift to something more practical, based on our spaceflights-pandas starter. I previously kept the cooking example, but let's go ahead and remove it.

yury-fedotov commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -252,3 +252,8 @@ Nodes that are created with input or output names that contain `.` risk a discon
+
+## How to store pipeline code in a kedro project
+
+When managing your Kedro project, we recommend grouping related tasks into individual pipelines to achieve modularity. A project typically contains many tasks, and organising frequently executed tasks together into separate pipelines helps maintain order and efficiency. Each pipeline should ideally be organised in its own folder, promoting easy copying and reuse within and between projects. Simpl ...
```
Probably not `between projects` because of micro packaging deprecation?

yury-fedotov commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,336 @@
+from kedro.pipeline import Pipeline, pipeline
+from ..data_science.pipeline import base_data_science  # Import pipeline to create a new one based on it
+
+def create_pipeline(**kwargs) -> Pipeline:
```
```suggestion
def create_pipeline() -> Pipeline:
```

yury-fedotov reviewed (commented):
Hey! I read through this a bit, if I may, would leave a couple of comments.

> I definitely don't have as deep knowledge on namespaces and modular pipelines as maintainers, but I'd say I leverage them comfortably and e.g. making dynamic namespaced pipelines (like described [here](https://getindata.com/blog/kedro-dynamic-pipelines/)) and nesting namespaces 2-3 levels deep is something I operate with very frequently. I even [found a bug](https://github.com/kedro-org/kedro-viz/issues/1814) in `kedro viz` at some point related to this :)

Overall I like the [current docs](https://docs.kedro.org/en/stable/nodes_and_pipelines/modular_pipelines.html) a lot more. It is subjective and personal opinion indeed, but a few things it's based on:

1. See multiple comments I left.
2. Modular pipelines and namespaces are IMO very interconnected and at least for me it's easy to follow if they're on the same page.

astrojuanlu commented:
Hi @yury-fedotov, thanks a lot for your review! 🙏🏼 

I'd like to add clarification on two bits though:

- The explanation on the wrapper was a _lie_: `kedro.pipeline.modular_pipeline.pipeline` and `kedro.pipeline.pipeline` are exactly the same, as I painfully discovered in #2723 one year ago when I taught namespaces for the first time. I'd even advocate to make it private API, but that's for another day. Regardless, it's already time to remove it from our docs.
- I don't think modular pipelines and namespaces are connected at all. In fact, most Kedro projects make use of modular pipelines (since that's the output of `kedro pipeline create`) but not all of them use namespaces, and it's fine. What's more: somebody could have non-modular pipelines (hence by _not_ following the directory structure recommended by Kedro) and still use namespaces. The messy terminology made these concepts difficult to understand even for Kedro power users and the Kedro team itself (in fact, this PR was prompted by [a discussion about `kedro-airflow`](https://github.com/kedro-org/kedro/issues/1998#issuecomment-2115696023)), so we considered it was urgent to clarify all these topics.

I'll let @DimedS act on the rest of your review comments as he sees fit.

yury-fedotov commented:
@astrojuanlu 

To be on the same page in terminology, by modular pipeline I mean a pipeline that is:

1. Defined but not registered itself
2. Leveraged in `pipeline()` wrapper, which overrides abstract inputs and outputs by specific catalog datasets, and this let's say "instance" is what's being registred.

> most Kedro projects make use of modular pipelines

Yeah but I'd assume most use them in a non-modular fashion, i.e. not leveraging `inputs`, `outputs`, `parameters` and `namespace` arguments of `pipeline`. Actually because of this what I do is after `kedro pipeline create` I go to `pipeline.py` and change the pipeline constructor from `pipeline` to `Pipeline` if I do not use namespaces.

> I don't think modular pipelines and namespaces are connected at all

I'm curious to understand why. IMO there are 2 reasons in Kedro to use modular pipelines:

1. Simplify `viz` output and overall mental model by namespacing all internals so that other pipelines care only about what's on the edges.
2. Reuse the same pipeline structure more than once.

In either case, `namespace` is what enables you do that. I don't understand why are they not connected.

Can you give an example of when you'd leverage a modular pipeline without a `namespace`?

astrojuanlu commented:
> by modular pipeline I mean a pipeline that is:

I don't claim to be in possession of the right meaning for "modular pipeline" but we are talking about 2 different things. Maybe we need some arbitration :) @idanov @merelcht @noklam

yury-fedotov commented:
> > by modular pipeline I mean a pipeline that is:
> 
> I don't claim to be in possession of the right meaning for "modular pipeline" but we are talking about 2 different things. Maybe we need some arbitration :) @idanov @merelcht @noklam

Sure, I didn't mean my definition is correct, that's just what I have in mind currently. Would be happy to change my mind if it's inaccurate.

But overall, when you say that modular pipelines and namespaces aren't connected, is there an example of a pipeline that's modular but not using namespaces to act as modular?

astrojuanlu commented:
> is there an example of a pipeline that's modular but not using namespaces to act as modular?

According to my current understanding of what modular pipelines are,

## Modular with no namespaces

That one is easy:

```
$ kedro pipeline create
```

...

```python
def create_pipeline(**kwargs) -> Pipeline:
    return pipeline(
        [
            node(
                func=preprocess_companies,
                inputs="companies",
                outputs="preprocessed_companies",
                name="preprocess_companies_node",
            ),
            node(
                func=preprocess_shuttles,
                inputs="shuttles",
                outputs="preprocessed_shuttles",
                name="preprocess_shuttles_node",
            ),
            node(
                func=create_model_input_table,
                inputs=["preprocessed_shuttles", "preprocessed_companies", "reviews"],
                outputs="model_input_table",
                name="create_model_input_table_node",
            ),
        ]
    )
```

## Not modular with namespaces

In `pipeline_registry.py`:

```python
def register_pipelines() -> Dict[str, Pipeline]:
    base_pipeline = pipeline(
        [
            node(
                func=split_data,
                inputs=["model_input_table", "params:model_options"],
                outputs=["X_train", "X_test", "y_train", "y_test"],
                name="split_data_node",
            ),
            node(
                func=train_model,
                inputs=["X_train", "y_train"],
                outputs="regressor",
                name="train_model_node",
            ),
            node(
                func=evaluate_model,
                inputs=["regressor", "X_test", "y_test"],
                outputs=None,
                name="evaluate_model_node",
            ),
        ]
    )
    ds_pipeline_1 = pipeline(
        pipe=base_pipeline,
        inputs="model_input_table",
        namespace="active_modelling_pipeline",
    )
    ds_pipeline_2 = pipeline(
        pipe=base_pipeline,
        inputs="model_input_table",
        namespace="candidate_modelling_pipeline",
    )

    return {"__default__": ds_pipeline_1 + ds_pipeline_2}
```

So the pipeline is defined directly in the `pipeline_registry.py`, which is perhaps a bad practice but nonetheless allowed, and it makes use of namespaces.

astrojuanlu commented:
Also note that our docs already say that the way to create & delete modular pipelines is with `kedro pipeline create/delete` https://docs.kedro.org/en/0.19.6/development/commands_reference.html#modular-pipelines

yury-fedotov commented:
@astrojuanlu thanks for providing those examples!

Yeah I think we just define modular pipelines differently. To simplify context a bit, let's assume that `kedro pipeline create/delete` do not exist, and we create pipeline folder structure ourselves. It actually takes a few seconds to do, so I think this example is quite realistic.

- Your first examples (with preprocessing companies and shuttles) - in what way is it modular? It points to specific items in the catalog, hence cannot be reused multiple times with different inputs and outputs. Well, in theory it can of course, but how is it different from the case where I'd just use a `Pipeline()` constructor as give it a list of nodes?
- What you have as a `base_pipeline` is what I would call modular. It is reusable and not registred itself, but rather 2 instances of that are created. They are created using **namespaces**, which allow reusability. I get why you call it not modular, but this is just because it's defined directly in the `pipeline_registry.py`, which to me is a weird pattern. If we take all this:

```python
base_pipeline = pipeline(
        [
            node(
                func=split_data,
                inputs=["model_input_table", "params:model_options"],
                outputs=["X_train", "X_test", "y_train", "y_test"],
                name="split_data_node",
            ),
            node(
                func=train_model,
                inputs=["X_train", "y_train"],
                outputs="regressor",
                name="train_model_node",
            ),
            node(
                func=evaluate_model,
                inputs=["regressor", "X_test", "y_test"],
                outputs=None,
                name="evaluate_model_node",
            ),
        ]
    )
    ds_pipeline_1 = pipeline(
        pipe=base_pipeline,
        inputs="model_input_table",
        namespace="active_modelling_pipeline",
    )
    ds_pipeline_2 = pipeline(
        pipe=base_pipeline,
        inputs="model_input_table",
        namespace="candidate_modelling_pipeline",
    )
```

Move it to manually created `pipelines/data_science/pipeline.py`, and then import `ds_pipeline_1` and `ds_pipeline_2` in registry, and registred. Would this make this pipeline modular?

stichbury commented:
I'm happy to do a final read through when further requested changes are complete. Please just ping me @DimedS when you're ready!

DimedS commented:
Many thanks for your review, @yury-fedotov, and your comments, @astrojuanlu. I mostly agree with @astrojuanlu's understanding that modular pipelines are currently a more virtual concept. For me, clarifying exactly what that means is not as important. I actually wanted to remove the term "modular" entirely to not to confuse users. What's more important to me is to clarify for users how to effectively use Kedro to solve their different tasks.

I like the current docs structure from that perspective because it's more task-oriented now:

1. In the intro, we explain that a pipeline is a combination of nodes with auto-execution order resolution.
2. In the modular pipelines section, we discuss the benefits of storing pipelines in separate folders, teach how to use the `pipeline create` command, and describe the general pipeline structure.
3. In the last and most complicated section, we explain how to reuse pipelines with namespaces. This is definitely an advanced-level section, and I hope it is now clearer for beginners.
@yury-fedotov, I will address your other comments today.

DimedS commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,345 @@
+    )
+```
+
+To use a new set of parameters, we need to previously save them inside `conf/base/parameters.yml` of our kedro project. I copied `model_options` from `conf/base/parameters_data_science.yml` and slightly modified it to try new model training parameters: test size and different features set, and called it `model_options_1`:
````
In my opinion, the point of the first section titled "How to reuse your pipelines" is to show that you can reuse your pipelines without needing namespaces. However, by adding namespaces, you can reuse your pipelines multiple times, with namespaces providing the necessary support. So my point is: reuse is the primary idea, and namespaces are an enhancement. I propose to leave it as it is.

DimedS commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,345 @@
+    - company_rating
+```
+
+> **Warning**: In Kedro, you cannot run pipelines with the same node names. In this example, both pipelines have nodes with the same names, so it's impossible to execute them together. However, `base_data_science` is not registered and will not be executed with the `kedro run` command. The `data_science` pipeline created inside the `create_pipeline()` function, based on `base_data_science`, will  ...
````
I rewrote it slightly, but I think it's important to mention that part because it explains why you actually need namespaces—because you can reuse pipelines without them.

merelcht commented on docs/source/nodes_and_pipelines/namespaces.md:
````diff
@@ -0,0 +1,166 @@
+    - company_rating
+```
+
+> In Kedro, you cannot run pipelines with the same node names. In this example, both pipelines have nodes with the same names, so it's impossible to execute them together. However, `base_data_science` is not registered and will not be executed with the `kedro run` command. The `data_science` pipeline, on the other hand, will be executed during `kedro run` because it will be autodiscovered by Kedr ...
````
Other than in the explanation there's no actual reference to `data_science` anywhere in the code examples on this page.

DimedS replied:
I fixed it by adding the comment `# data_science pipeline creation function` before the `create_pipeline()` function. However, as I understand it, I have to use `create_pipeline()` name to ensure the pipeline will be autodiscovered. Do you have other ideas on how to highlight the `data_science` name inside the code?

merelcht commented on docs/source/nodes_and_pipelines/namespaces.md:
```diff
@@ -0,0 +1,166 @@
+
+![namespaces collapsed](../meta/images/namespaces_collapsed.png)
+
+> Tip: You can use `kedro run --namespace = namespace_name` to run only the specific namespace
```
```suggestion
> Tip: You can use `kedro run --namespace=namespace_name` to run only the specific namespace
```

merelcht commented on docs/source/nodes_and_pipelines/pipeline_introduction.md:
```diff
@@ -1,26 +1,38 @@
 We previously introduced [Nodes](./nodes.md) as building blocks that represent tasks, and can be combined in a pipeline to build your workflow. A pipeline organises the dependencies and execution order of your collection of nodes, and connects inputs and outputs while keeping your code modular. The pipeline resolves dependencies to determine the node execution order, and does *not* necessarily ru ...
 
-To benefit from Kedro's automatic dependency resolution, you can chain your nodes into a {py:class}`~kedro.pipeline.Pipeline`, which is a list of nodes that use a shared set of variables.
+To benefit from Kedro's automatic dependency resolution, you can chain your nodes into a {py:class}`~kedro.pipeline.Pipeline`, which is a list of nodes that use a shared set of variables. That class can be created using the {py:class}`~kedro.pipeline.modular_pipeline.pipeline` method, based on nodes or other pipelines (in which case all nodes from that pipeline will be used).
```
```suggestion
To benefit from Kedro's automatic dependency resolution is that you can chain your nodes into a {py:class}`~kedro.pipeline.Pipeline`, which is a list of nodes that use a shared set of variables. That class can be created using the {py:class}`~kedro.pipeline.modular_pipeline.pipeline` method, based on nodes or other pipelines (in which case all nodes from that pipeline will be used).
```

merelcht reviewed (approved):
Some small final suggestions, but nothing blocking. 

Thanks for the many revisions of these docs @DimedS ⭐ From my point of view this is now a lot better than what we had before. 

I would like us to put some effort in creating a more realistic example of re-using the pipeline so instead of data_science_1 and data_science_2 we can refer to a proper use case that requires re-use, but let's leave that for another time.

stichbury commented:
> I'm happy to do a final read through when further requested changes are complete. Please just ping me @DimedS when you're ready!

Oh, you merged it. I can still see quite a few Vale issues that I would have preferred to have been resolved, but I guess they can be addressed separately as needed.

astrojuanlu commented:
Adding that as a follow-up task in https://github.com/kedro-org/kedro/issues/2723#issuecomment-2202784458

DimedS commented:
> > I'm happy to do a final read through when further requested changes are complete. Please just ping me @DimedS when you're ready!
> 
> Oh, you merged it. I can still see quite a few Vale issues that I would have preferred to have been resolved, but I guess they can be addressed separately as needed.

I'm sorry @stichbury, I forgot to ping you before merging. There were many comments in the PR, and we agreed to continue addressing part of them in another PR. Thanks for the follow-up task description, @astrojuanlu.
